diff --git a/conf/battle/monster.conf b/conf/battle/monster.conf
index 939da1fd..6fa53d6e 100644
--- a/conf/battle/monster.conf
+++ b/conf/battle/monster.conf
@@ -332,6 +332,25 @@ achievement_mob_share: no
 // Default (Official): 1000
 mob_respawn_time: 1000
 
+// [RomuloSM]: Mob MvP Effect
+// Show MvP Effect in monsters?
+mob_show_mvp_effect: yes
+
+// Refresh Timer for MvP Effect. (in milliseconds)
+mob_show_mvp_effect_timer: 2000
+
+// [RomuloSM]: Mob Hat Effects
+// Show hat effect of monster Elements?
+mob_show_hateffect_element: yes
+
+// Show hat effect of monster Race?
+mob_show_hateffect_race: yes
+
+// [RomuloSM]: Mob Hat Effects
+// Show hat effect of monster Quest?
+mob_show_hateffect_quest: yes
+
+
 // Defines the time (in ms) after which a monster unlocks its target when trapped.
 // During this time monsters will still be in idle mode and use idle skills on random
 // targets, but they continue chasing their original target when no longer trapped.
diff --git a/src/map/atcommand.cpp b/src/map/atcommand.cpp
index 8165ef24..a652154d 100644
--- a/src/map/atcommand.cpp
+++ b/src/map/atcommand.cpp
@@ -11414,6 +11414,63 @@ ACMD_FUNC(macrochecker){
 
 #include <custom/atcommand.inc>
 
+// [RomuloSM]: Show Mob MvP Effect
+ACMD_FUNC(showmobmvp) {
+	nullpo_retr(-1, sd);
+
+	if( sd->showMobMvPEffect ) {
+		sd->showMobMvPEffect = false;
+		clif_displaymessage(fd, "Show Mobs MvP: On");
+	} else {
+		sd->showMobMvPEffect = true;
+		clif_displaymessage(fd, "Show Mobs MvP: Off");
+	}
+	return 0;
+}
+
+// [RomuloSM]: Show Mobs Hat Effects
+ACMD_FUNC(showmobelement) {
+	nullpo_retr(-1, sd);
+
+	if( sd->showMobHatEffectElement ) {
+		sd->showMobHatEffectElement = false;
+		clif_displaymessage(fd, "Show Mobs Element: On");
+	} else {
+		sd->showMobHatEffectElement = true;
+		clif_displaymessage(fd, "Show Mobs Element: Off");
+	}
+	map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+	return 0;
+}
+
+ACMD_FUNC(showmobrace) {
+	nullpo_retr(-1, sd);
+
+	if( sd->showMobHatEffectRace ) {
+		sd->showMobHatEffectRace = false;
+		clif_displaymessage(fd, "Show Mobs Race: On");
+	} else {
+		sd->showMobHatEffectRace = true;
+		clif_displaymessage(fd, "Show Mobs Race: Off");
+	}
+	map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+	return 0;
+}
+
+ACMD_FUNC(showmobquest) {
+	nullpo_retr(-1, sd);
+
+	if( sd->showMobHatEffectQuest ) {
+		sd->showMobHatEffectQuest = false;
+		clif_displaymessage(fd, "Show Mobs Quest: On");
+	} else {
+		sd->showMobHatEffectQuest = true;
+		clif_displaymessage(fd, "Show Mobs Quest: Off");
+	}
+	map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+	return 0;
+}
+
 /**
  * Fills the reference of available commands in atcommand DBMap
  **/
@@ -11741,6 +11798,12 @@ void atcommand_basecommands(void) {
 		ACMD_DEFR(enchantgradeui, ATCMD_NOCONSOLE|ATCMD_NOAUTOTRADE),
 		ACMD_DEFR(roulette, ATCMD_NOCONSOLE|ATCMD_NOAUTOTRADE),
 		ACMD_DEF(setcard),
+		// [RomuloSM]: Show Mob MvP Effect
+		ACMD_DEF(showmobmvp),
+		// [RomuloSM]: Show Mobs Hat Effects
+		ACMD_DEF(showmobelement),
+		ACMD_DEF(showmobrace),
+		ACMD_DEF(showmobquest),
 		ACMD_DEF(macrochecker),
 	};
 	AtCommandInfo* atcommand;
diff --git a/src/map/battle.cpp b/src/map/battle.cpp
index 922cc872..f22d99e3 100644
--- a/src/map/battle.cpp
+++ b/src/map/battle.cpp
@@ -12363,6 +12363,15 @@ static const struct _battle_data {
 	{ "trade_count_stackable",              &battle_config.trade_count_stackable,           1,      0,      1,              },
 	{ "enable_bonus_map_drops",             &battle_config.enable_bonus_map_drops,          1,      0,      1,              },
 
+	// [RomuloSM]: Mob MvP Effect
+	{ "mob_show_mvp_effect",                &battle_config.mob_show_mvp_effect,             1,      0,      1,              },
+	{ "mob_show_mvp_effect_timer",          &battle_config.mob_show_mvp_effect_timer,    1000,   2000,      INT_MAX,        },
+
+	// [RomuloSM]: Mob Hat Effect Element
+	{ "mob_show_hateffect_element",         &battle_config.mob_show_hateffect_element,      1,      0,      1,              },
+	{ "mob_show_hateffect_race",            &battle_config.mob_show_hateffect_race,         1,      0,      1,              },
+	{ "mob_show_hateffect_quest",           &battle_config.mob_show_hateffect_quest,        1,      0,      1,              },
+
 #include <custom/battle_config_init.inc>
 };
 
diff --git a/src/map/battle.hpp b/src/map/battle.hpp
index 845cd5fa..5ef363ea 100644
--- a/src/map/battle.hpp
+++ b/src/map/battle.hpp
@@ -782,6 +782,15 @@ struct Battle_Config
 	int32 trade_count_stackable;
 	int32 enable_bonus_map_drops;
 
+	// [RomuloSM]: Mob MvP effect
+	int mob_show_mvp_effect;
+	int mob_show_mvp_effect_timer;
+
+	// [RomuloSM]: Mob Hat Effect Element
+	int mob_show_hateffect_element;
+	int mob_show_hateffect_race;
+	int mob_show_hateffect_quest;
+
 #include <custom/battle_config_struct.inc>
 };
 
diff --git a/src/map/clif.cpp b/src/map/clif.cpp
index afbdd588..25750629 100644
--- a/src/map/clif.cpp
+++ b/src/map/clif.cpp
@@ -1718,6 +1718,8 @@ int32 clif_spawn( block_list *bl, bool walking ){
 				clif_specialeffect(md,EF_BABYBODY2,AREA);
 			if ( md->special_state.ai == AI_ABR || md->special_state.ai == AI_BIONIC )
 				clif_summon_init(*md);
+			// [RomuloSM]: Mob Hat Effects
+			map_foreachinallrange(mob_hateffect_sub, &md->bl, AREA_SIZE, BL_PC, md);
 			clif_name_area(md);
 		}
 		break;
@@ -25727,6 +25729,8 @@ static int32 clif_parse(int32 fd)
 				ShowWarning( "clif_parse: It looks like you have disabled PACKET_OBFUSCATION on server side, but enabled it on client side.\n" );
 			}
 #endif
+			// [RomuloSM]: Mob Hat Effects
+			clif_mob_hat_effects(md,&sd->bl, SELF);
 		}
 #endif
 
@@ -25832,6 +25836,102 @@ void packetdb_readdb(){
 #endif
 }
 
+/// [RomuloSM]: Mob Hat Effects
+/// Send hat effects to the client (ZC_HAT_EFFECT).
+/// 0A3B <Length>.W <AID>.L <Status>.B { <HatEffectId>.W }
+void clif_mob_hat_effects( struct mob_data* md, struct block_list* bl, enum send_target target ){
+#if PACKETVER_MAIN_NUM >= 20150507 || PACKETVER_RE_NUM >= 20150429 || defined(PACKETVER_ZERO)
+	nullpo_retv( md );
+	nullpo_retv( bl );
+
+	if( bl->type != BL_PC || map_getmapdata(bl->m)->getMapFlag(MF_NOCOSTUME) ){
+		return;
+	}
+
+	struct PACKET_ZC_EQUIPMENT_EFFECT* p = (struct PACKET_ZC_EQUIPMENT_EFFECT*)packet_buffer;
+	size_t i = 0, pSize = md->hatEffects.size();
+
+	if( !md->hatEffects.empty() ) {
+		for( i = 0; i < md->hatEffects.size(); i++ ){
+			p->effects[i] = md->hatEffects[i];
+		}
+	}
+
+	int hub = 1;
+	int64 id;
+	map_session_data* sd = BL_CAST( BL_PC, bl );
+	if( battle_config.mob_show_hateffect_element && sd && (id = md->get_hatelement(hub)) != HAT_EF_MIN && !sd->showMobHatEffectElement ) {
+		p->effects[i] = static_cast<e_hat_effects>(id);
+		i++;
+		hub++;
+	}
+
+	if( battle_config.mob_show_hateffect_race && sd && (id = md->get_hatrace(hub)) != HAT_EF_MIN && !sd->showMobHatEffectRace ) {
+		p->effects[i] = static_cast<e_hat_effects>(id);
+		i++;
+		hub++;
+	}
+
+	if( battle_config.mob_show_hateffect_quest && sd && pc_mob_quest_check(sd,md->mob_id) && !sd->showMobHatEffectQuest ) {
+		std::string constant = "HAT_EF_QUEST_OBJECTIVE_" + std::to_string(hub);
+		if( script_get_constant(constant.c_str(), &id) ) {
+			p->effects[i] = static_cast<e_hat_effects>(id);
+			i++;
+			hub++;
+		}
+	}
+
+	if( !i ) {
+		return;
+	}
+
+	p->packetType = HEADER_ZC_EQUIPMENT_EFFECT;
+	p->packetLength = (int16)(sizeof(struct PACKET_ZC_EQUIPMENT_EFFECT) + sizeof(int16) * i);
+	p->aid = md->bl.id;
+	p->status = 1;
+
+	clif_send( p, p->packetLength, bl, target );
+#endif
+}
+
+void clif_mob_hat_effect_single( struct mob_data *md, map_session_data* sd, uint16 effectId, bool enable ){
+#if PACKETVER_MAIN_NUM >= 20150507 || PACKETVER_RE_NUM >= 20150429 || defined(PACKETVER_ZERO)
+	nullpo_retv( md );
+	nullpo_retv( sd );
+
+	struct PACKET_ZC_EQUIPMENT_EFFECT* p = (struct PACKET_ZC_EQUIPMENT_EFFECT*)packet_buffer;
+
+	p->packetType = HEADER_ZC_EQUIPMENT_EFFECT;
+	p->packetLength = (int16)( sizeof( struct PACKET_ZC_EQUIPMENT_EFFECT ) + sizeof( int16 ) );
+	p->aid = md->bl.id;
+	p->status = enable;
+	p->effects[0] = effectId;
+
+	clif_send( p, p->packetLength, &sd->bl, SELF );
+#endif
+}
+
+void clif_mob_hat_effect_hub_remove( struct mob_data *md, map_session_data* sd ){
+#if PACKETVER_MAIN_NUM >= 20150507 || PACKETVER_RE_NUM >= 20150429 || defined(PACKETVER_ZERO)
+	nullpo_retv( md );
+	nullpo_retv( sd );
+
+	int i;
+
+	// Elements
+	for (i = HAT_EF_ELE_NEUTRAL_1; i <= HAT_EF_ELE_UNDEAD_3; i++)
+		clif_mob_hat_effect_single(md, sd, i, false);
+
+	// Race
+	for (i = HAT_EF_RC_FORMLESS_1; i <= HAT_EF_RC_PLAYER_DORAM_3; i++)
+		clif_mob_hat_effect_single(md, sd, i, false);
+
+	// Quests
+	for (i = HAT_EF_QUEST_OBJECTIVE_1; i <= HAT_EF_QUEST_OBJECTIVE_3; i++)
+		clif_mob_hat_effect_single(md, sd, i, false);
+#endif
+}
+
 /*==========================================
  *
  *------------------------------------------*/
diff --git a/src/map/clif.hpp b/src/map/clif.hpp
index 535bb22c..b0013222 100644
--- a/src/map/clif.hpp
+++ b/src/map/clif.hpp
@@ -1509,4 +1509,9 @@ void clif_noask_sub( map_session_data& sd, map_session_data& tsd, int32 type );
 
 void clif_specialpopup(map_session_data& sd, int32 id);
 
+/// [RomuloSM]: Mob Hat Effects
+void clif_mob_hat_effects( struct mob_data* md, struct block_list* bl, enum send_target target );
+void clif_mob_hat_effect_single( struct mob_data *md, map_session_data* sd, uint16 effectId, bool enable );
+void clif_mob_hat_effect_hub_remove( struct mob_data *md, map_session_data* sd );
+
 #endif /* CLIF_HPP */
diff --git a/src/map/mob.cpp b/src/map/mob.cpp
index 1b494059..d216c4f2 100644
--- a/src/map/mob.cpp
+++ b/src/map/mob.cpp
@@ -394,6 +394,109 @@ e_mob_bosstype mob_data::get_bosstype(){
 	}
 }
 
+// [RomuloSM]: Mob Hat Effects
+e_hat_effects mob_data::get_hatelement(int hub) {
+	if( this->db != nullptr ){
+		std::string name;
+		switch( this->db->status.def_ele ) {
+			case ELE_NEUTRAL:
+				name = "NEUTRAL";
+				break;
+			case ELE_WATER:
+				name = "WATER";
+				break;
+			case ELE_EARTH:
+				name = "EARTH";
+				break;
+			case ELE_FIRE:
+				name = "FIRE";
+				break;
+			case ELE_WIND:
+				name = "WIND";
+				break;
+			case ELE_POISON:
+				name = "POISON";
+				break;
+			case ELE_HOLY:
+				name = "HOLY";
+				break;
+			case ELE_DARK:
+				name = "DARK";
+				break;
+			case ELE_GHOST:
+				name = "GHOST";
+				break;
+			case ELE_UNDEAD:
+				name = "UNDEAD";
+				break;
+			default:
+				return HAT_EF_MIN;
+		}
+
+		int64 id;
+		if( !script_get_constant(("HAT_EF_ELE_" + name + "_" + std::to_string(hub)).c_str(), &id) )
+			return HAT_EF_MIN;
+
+		return static_cast<e_hat_effects>(id);
+	}else{
+		return HAT_EF_MIN;
+	}
+}
+
+e_hat_effects mob_data::get_hatrace(int hub) {
+	if( this->db != nullptr ){
+		std::string name;
+		switch( this->db->status.race ) {
+			case RC_FORMLESS:
+				name = "FORMLESS";
+				break;
+			case RC_UNDEAD:
+				name = "UNDEAD";
+				break;
+			case RC_BRUTE:
+				name = "BRUTE";
+				break;
+			case RC_PLANT:
+				name = "PLANT";
+				break;
+			case RC_INSECT:
+				name = "INSECT";
+				break;
+			case RC_FISH:
+				name = "FISH";
+				break;
+			case RC_DEMON:
+				name = "DEMON";
+				break;
+			case RC_DEMIHUMAN:
+				name = "DEMIHUMAN";
+				break;
+			case RC_ANGEL:
+				name = "ANGEL";
+				break;
+			case RC_DRAGON:
+				name = "DRAGON";
+				break;
+			case RC_PLAYER_HUMAN:
+				name = "PLAYER_HUMAN";
+				break;
+			case RC_PLAYER_DORAM:
+				name = "PLAYER_DORAM";
+				break;
+			default:
+				return HAT_EF_MIN;
+		}
+
+		int64 id;
+		if( !script_get_constant(("HAT_EF_RC_" + name + "_" + std::to_string(hub)).c_str(), &id) )
+			return HAT_EF_MIN;
+
+		return static_cast<e_hat_effects>(id);
+	}else{
+		return HAT_EF_MIN;
+	}
+}
+
 /**
  * Create unique view data associated to a spawned monster.
  * @param md: Mob to adjust
@@ -1220,6 +1323,12 @@ int32 mob_spawn (mob_data *md)
 		clif_spawn(md);
 	skill_unit_move(md,tick,1);
 	mobskill_use(md, tick, MSC_SPAWN);
+
+	// [RomuloSM]: Mob MvP Effect
+	md->mvp_effect_tid = INVALID_TIMER;
+	if( md->get_bosstype() == BOSSTYPE_MVP )
+		md->mvp_effect_tid = add_timer(gettick() + 1, mob_mvp_effect_timer, md->bl.id, 0);
+
 	return 0;
 }
 
@@ -3604,6 +3713,12 @@ int32 mob_dead(mob_data *md, block_list *src, int32 type)
 		}
 	}
 
+	// [RomuloSM]: Mob Mvp Effect
+	if(md->mvp_effect_tid != INVALID_TIMER) {
+		delete_timer(md->mvp_effect_tid,mob_mvp_effect_timer);
+		md->mvp_effect_tid = INVALID_TIMER;
+	}
+
 	if(md->deletetimer != INVALID_TIMER) {
 		delete_timer(md->deletetimer,mob_timer_delete);
 		md->deletetimer = INVALID_TIMER;
@@ -3643,6 +3758,10 @@ int32 mob_dead(mob_data *md, block_list *src, int32 type)
 	if (battle_config.mvp_tomb_enabled && md->spawn->state.boss && map_getmapflag(md->m, MF_NOTOMB) != 1)
 		mvptomb_create(md, mvp_sd != nullptr ? mvp_sd->status.name : (first_sd != nullptr ? first_sd->status.name : nullptr), time(nullptr));
 
+	// [RomuloSM]: Mob Hat Effects
+	if( battle_config.mob_show_hateffect_quest )
+		map_foreachinallrange(mob_hateffect_sub, &md->bl, AREA_SIZE, BL_PC, md);
+
 	if( !rebirth )
 		mob_setdelayspawn(md); //Set respawning.
 	return 3; //Remove from map.
@@ -4799,6 +4918,66 @@ int32 mob_clone_delete(mob_data *md){
 	return 0;
 }
 
+// [RomuloSM]: Mob MvP Effect
+int mob_mvp_sub(struct block_list *bl, va_list ap) {
+	struct mob_data *md;
+	map_session_data *sd;
+
+	sd = (map_session_data *)bl;
+	md = va_arg(ap,mob_data *);
+	
+	nullpo_ret(sd);
+	nullpo_ret(md);
+
+	if( sd && md ) {
+		if( battle_config.mob_show_mvp_effect && !sd->showMobMvPEffect )
+			clif_specialeffect_single(&md->bl,EF_MVP,sd->fd);
+	}
+	return 1;
+}
+
+TIMER_FUNC(mob_mvp_effect_timer) {
+	struct block_list* bl = map_id2bl(id);
+	struct mob_data* md = BL_CAST(BL_MOB, bl);
+
+	if( !md )
+		return 0;
+
+	if( md->mvp_effect_tid != tid ) {
+		ShowError("mob_mvp_effect_timer: Timer mismatch: %d != %d\n", tid, md->mvp_effect_tid);
+		return 0;
+	}
+
+	if( md->mvp_effect_tid != INVALID_TIMER ) {
+		delete_timer(md->mvp_effect_tid, mob_mvp_effect_timer);
+		md->mvp_effect_tid = INVALID_TIMER;
+	}
+
+	if( battle_config.mob_show_mvp_effect )
+		map_foreachinallrange(mob_mvp_sub, &md->bl, AREA_SIZE, BL_PC, md);
+	md->mvp_effect_tid = add_timer(gettick() + battle_config.mob_show_mvp_effect_timer, mob_mvp_effect_timer, md->bl.id, 0);
+	return 1;
+}
+
+// [RomuloSM]: Mob Hat Effects
+int mob_hateffect_sub(struct block_list *bl, va_list ap)
+{
+	struct mob_data *md;
+	map_session_data *sd;
+
+	sd = (map_session_data *)bl;
+	md = va_arg(ap,mob_data *);
+	
+	nullpo_ret(sd);
+	nullpo_ret(md);
+
+	if( sd && md ) {
+		clif_mob_hat_effect_hub_remove(md,sd);
+		clif_mob_hat_effects(md, &sd->bl, SELF);
+	}
+	return 1;
+}
+
 //Adjusts the drop rate of item according to the criteria given. [Skotlex]
 static uint32 mob_drop_adjust(int32 baserate, int32 rate_adjust, uint16 rate_min, uint16 rate_max)
 {
@@ -7356,6 +7535,9 @@ void do_init_mob(void){
 	add_timer_func_list(mob_norm_attacked, "mob_norm_attacked");
 	add_timer_interval(gettick()+MIN_MOBTHINKTIME,mob_ai_hard,0,0,MIN_MOBTHINKTIME);
 	add_timer_interval(gettick()+MIN_MOBTHINKTIME*10,mob_ai_lazy,0,0,MIN_MOBTHINKTIME*10);
+
+	// [RomuloSM]: Mob MvP Effect
+	add_timer_func_list(mob_mvp_effect_timer,"mob_mvp_effect_timer");
 }
 
 /*==========================================
diff --git a/src/map/mob.hpp b/src/map/mob.hpp
index 63f30207..920ffe87 100644
--- a/src/map/mob.hpp
+++ b/src/map/mob.hpp
@@ -281,6 +281,16 @@ struct s_mob_db {
 	std::string title;
 
 	e_mob_bosstype get_bosstype();
+
+	// [RomuloSM]: Mob MvP Effect
+	int mvp_effect_tid;
+
+	// [RomuloSM]: Mob Hat Effects
+#if PACKETVER_MAIN_NUM >= 20150507 || PACKETVER_RE_NUM >= 20150429 || defined(PACKETVER_ZERO)
+	std::vector<int16> hatEffects;
+#endif
+	e_hat_effects get_hatelement(int hub);
+	e_hat_effects get_hatrace(int hub);
 	s_mob_db();
 };
 
@@ -399,6 +409,16 @@ struct mob_data : public block_list {
 	uint16 damagetaken;
 
 	e_mob_bosstype get_bosstype();
+
+	// [RomuloSM]: Mob MvP Effect
+	int mvp_effect_tid;
+
+	// [RomuloSM]: Mob Hat Effects
+#if PACKETVER_MAIN_NUM >= 20150507 || PACKETVER_RE_NUM >= 20150429 || defined(PACKETVER_ZERO)
+	std::vector<int16> hatEffects;
+#endif
+	e_hat_effects get_hatelement(int hub);
+	e_hat_effects get_hatrace(int hub);
 	map_session_data* get_mvp_player();
 };
 
@@ -584,4 +604,11 @@ void mob_setdropitem_option( item& itm, const std::shared_ptr<s_mob_drop>& mobdr
 
 #define CHK_MOBSIZE(size) ((size) >= SZ_SMALL && (size) < SZ_MAX) /// Check valid Monster Size
 
+// [RomuloSM]: Mob MvP Effect
+int mob_mvp_sub(struct block_list *bl, va_list ap);
+TIMER_FUNC(mob_mvp_effect_timer);
+
+// [RomuloSM]: Mob Hat Effects
+int mob_hateffect_sub(struct block_list *bl, va_list ap);
+
 #endif /* MOB_HPP */
diff --git a/src/map/pc.cpp b/src/map/pc.cpp
index 90a6e9ba..4ea0324d 100644
--- a/src/map/pc.cpp
+++ b/src/map/pc.cpp
@@ -2155,6 +2155,10 @@ bool pc_authok(map_session_data *sd, uint32 login_id2, time_t expiration_time, i
 	sd->regen.tick.hp = tick;
 	sd->regen.tick.sp = tick;
 
+	// [RomuloSM]: Show Mobs Hat Effects
+	sd->showMobHatEffectElement = false;
+	sd->showMobHatEffectQuest = false;
+
 	for(int32 i = 0; i < MAX_SPIRITBALL; i++)
 		sd->spirit_timer[i] = INVALID_TIMER;
 
@@ -16027,6 +16031,51 @@ uint64 CaptchaDatabase::parseBodyNode(const ryml::NodeRef &node) {
 	return 1;
 }
 
+// [RomuloSM]: Mob Hat Effects
+int pc_mob_hateffect_sub(struct block_list *bl, va_list ap)
+{
+	map_session_data *sd;
+	struct mob_data *md;
+
+	md = (struct mob_data *)bl;
+	sd = va_arg(ap,map_session_data *);
+	
+	nullpo_ret(sd);
+	nullpo_ret(md);
+
+	if( sd && md ) {
+		clif_mob_hat_effect_hub_remove(md,sd);
+		clif_mob_hat_effects(md, &sd->bl, SELF);
+	}
+	return 1;
+}
+
+bool pc_mob_quest_check(map_session_data *sd, int mob_id)
+{
+	nullpo_retr(false,sd);
+
+	if( !mob_id )
+		return false;
+
+	for( int i = 0; i < sd->avail_quests; i++ ) {
+		if( sd->quest_log[i].state == Q_COMPLETE ) // Skip complete quests
+			continue;
+
+		std::shared_ptr<s_quest_db> qi = quest_search(sd->quest_log[i].quest_id);
+		if( !qi )
+			continue;
+
+		for( int j = 0; j < qi->objectives.size(); j++ ) {
+			uint8 objective_check = 0;
+
+			if( qi->objectives[j]->mob_id == mob_id && sd->quest_log[i].count[j] < qi->objectives[j]->count )
+				return true;
+		}
+	}
+
+	return false;
+}
+
 /*==========================================
  * pc Init/Terminate
  *------------------------------------------*/
diff --git a/src/map/pc.hpp b/src/map/pc.hpp
index 83299c7f..89fc72a3 100644
--- a/src/map/pc.hpp
+++ b/src/map/pc.hpp
@@ -946,6 +946,14 @@ public:
 
 	std::vector<uint32> party_booking_requests;
 
+	// [RomuloSM]: Show Mob MvP Effect
+	bool showMobMvPEffect;
+
+	// [RomuloSM]: Show Mobs Hat Effects
+	bool showMobHatEffectElement;
+	bool showMobHatEffectRace;
+	bool showMobHatEffectQuest;
+
 	void update_look( _look look );
 };
 
@@ -1771,4 +1779,8 @@ void pc_macro_reporter_process(map_session_data &sd, int32 reporter_account_id =
 void pc_reputation_generate();
 #endif
 
+// [RomuloSM]: Mob Hat Effects
+int pc_mob_hateffect_sub(struct block_list *bl, va_list ap);
+bool pc_mob_quest_check(map_session_data *sd, int mob_id);
+
 #endif /* PC_HPP */
diff --git a/src/map/quest.cpp b/src/map/quest.cpp
index 94f673e9..4c023f96 100644
--- a/src/map/quest.cpp
+++ b/src/map/quest.cpp
@@ -623,6 +623,10 @@ int32 quest_add(map_session_data *sd, int32 quest_id)
 	if( save_settings&CHARSAVE_QUEST )
 		chrif_save(sd, CSAVE_NORMAL);
 
+	// [RomuloSM]: Mob Hat Effect Element
+	if( battle_config.mob_show_hateffect_quest )
+		map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+
 	return 0;
 }
 
@@ -673,6 +677,10 @@ int32 quest_change(map_session_data *sd, int32 qid1, int32 qid2)
 	if( save_settings&CHARSAVE_QUEST )
 		chrif_save(sd, CSAVE_NORMAL);
 
+	// [RomuloSM]: Mob Hat Effect Element
+	if( battle_config.mob_show_hateffect_quest )
+		map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+
 	return 0;
 }
 
@@ -712,6 +720,10 @@ int32 quest_delete(map_session_data *sd, int32 quest_id)
 	if( save_settings&CHARSAVE_QUEST )
 		chrif_save(sd, CSAVE_NORMAL);
 
+	// [RomuloSM]: Mob Hat Effect Element
+	if( battle_config.mob_show_hateffect_quest )
+		map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+
 	return 0;
 }
 
@@ -835,6 +847,10 @@ void quest_update_objective(map_session_data *sd, mob_data* md)
 		}
 	}
 	pc_show_questinfo(sd);
+
+	// [RomuloSM]: Mob Hat Effect Element
+	if( battle_config.mob_show_hateffect_quest )
+		map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
 }
 
 /**
@@ -878,6 +894,10 @@ int32 quest_update_status(map_session_data *sd, int32 quest_id, e_quest_state st
 	if (save_settings&CHARSAVE_QUEST)
 		chrif_save(sd, CSAVE_NORMAL);
 
+	// [RomuloSM]: Mob Hat Effect Element
+	if( battle_config.mob_show_hateffect_quest )
+		map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+
 	return 0;
 }
 
diff --git a/src/map/script.cpp b/src/map/script.cpp
index 13e63f0f..d87210d4 100644
--- a/src/map/script.cpp
+++ b/src/map/script.cpp
@@ -27955,6 +27955,132 @@ BUILDIN_FUNC(preg_match) {
 #endif
 }
 
+// [RomuloSM]: Mob Hat Effects
+BUILDIN_FUNC(mob_hateffect){
+#if PACKETVER_MAIN_NUM >= 20150507 || PACKETVER_RE_NUM >= 20150429 || defined(PACKETVER_ZERO)
+	struct block_list* bl;
+	int16 effectID;
+	bool enable;
+	TBL_MOB* md;
+
+	if( !script_rid2bl(2,bl) ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	if( (md = map_id2md(bl->id)) == NULL ) {
+		ShowError( "buildin_mob_hateffect: invalid unit id %d\n", bl->id );
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	effectID = script_getnum(st,3);
+	enable = script_getnum(st,4) ? true : false;
+
+	if( effectID <= HAT_EF_MIN || effectID >= HAT_EF_MAX ){
+		ShowError( "buildin_mob_hateffect: unsupported hat effect id %d\n", effectID );
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	auto it = util::vector_get( md->hatEffects, effectID );
+
+	if( enable ){
+		if( it != md->hatEffects.end() ){
+			return SCRIPT_CMD_SUCCESS;
+		}
+
+		md->hatEffects.push_back( effectID );
+	}else{
+		if( it == md->hatEffects.end() ){
+			return SCRIPT_CMD_SUCCESS;
+		}
+
+		util::vector_erase_if_exists( md->hatEffects, effectID );
+	}
+
+	clif_mob_hat_effects(md, bl, AREA);
+
+#endif
+	return SCRIPT_CMD_SUCCESS;
+}
+
+BUILDIN_FUNC(mob_showelement) {
+#if PACKETVER_MAIN_NUM >= 20150507 || PACKETVER_RE_NUM >= 20150429 || defined(PACKETVER_ZERO)
+	map_session_data* sd;
+	int flag = script_getnum(st,2);
+
+	if( script_hasdata(st, 3) ){
+		if (script_isint(st, 3))
+			script_charid2sd(3, sd);
+		else
+			script_nick2sd(3, sd);
+	}else{
+		script_rid2sd(sd);
+	}
+
+	if( !sd ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	sd->showMobHatEffectElement = flag ? false : true;
+	map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+	script_pushint(st,1);
+#endif
+	return SCRIPT_CMD_SUCCESS;
+}
+
+BUILDIN_FUNC(mob_showrace) {
+#if PACKETVER_MAIN_NUM >= 20150507 || PACKETVER_RE_NUM >= 20150429 || defined(PACKETVER_ZERO)
+	map_session_data* sd;
+	int flag = script_getnum(st,2);
+
+	if( script_hasdata(st, 3) ){
+		if (script_isint(st, 3))
+			script_charid2sd(3, sd);
+		else
+			script_nick2sd(3, sd);
+	}else{
+		script_rid2sd(sd);
+	}
+
+	if( !sd ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	sd->showMobHatEffectRace = flag ? false : true;
+	map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+	script_pushint(st,1);
+#endif
+	return SCRIPT_CMD_SUCCESS;
+}
+
+BUILDIN_FUNC(mob_showquest) {
+#if PACKETVER_MAIN_NUM >= 20150507 || PACKETVER_RE_NUM >= 20150429 || defined(PACKETVER_ZERO)
+	map_session_data* sd;
+	int flag = script_getnum(st,2);
+
+	if( script_hasdata(st, 3) ){
+		if (script_isint(st, 3))
+			script_charid2sd(3, sd);
+		else
+			script_nick2sd(3, sd);
+	}else{
+		script_rid2sd(sd);
+	}
+
+	if( !sd ) {
+		script_pushint(st, 0);
+		return SCRIPT_CMD_FAILURE;
+	}
+
+	sd->showMobHatEffectQuest = flag ? false : true;
+	map_foreachinallrange(pc_mob_hateffect_sub, &sd->bl, AREA_SIZE, BL_MOB, sd);
+	script_pushint(st,1);
+#endif
+	return SCRIPT_CMD_SUCCESS;
+}
+
 /// script command definitions
 /// for an explanation on args, see add_buildin_func
 struct script_function buildin_func[] = {
@@ -28685,6 +28811,12 @@ struct script_function buildin_func[] = {
 
 	BUILDIN_DEF( mesitemicon, "v??" ),
 
+	// [RomuloSM]: Mob Hat Effects
+	BUILDIN_DEF(mob_hateffect, "iii"),
+	BUILDIN_DEF(mob_showelement, "i?"),
+	BUILDIN_DEF(mob_showrace, "i?"),
+	BUILDIN_DEF(mob_showquest, "i?"),
+
 #include <custom/script_def.inc>
 
 	{nullptr,nullptr,nullptr},
diff --git a/src/map/script_constants.hpp b/src/map/script_constants.hpp
index 26466cd4..8b86802c 100644
--- a/src/map/script_constants.hpp
+++ b/src/map/script_constants.hpp
@@ -10888,6 +10888,79 @@
 	export_constant(GUILDINFO_MASTERID);
 	export_constant(GUILDINFO_MASTERNAME);
 
+	// [RomuloSM]: Elements
+	export_constant(HAT_EF_ELE_NEUTRAL_1);
+	export_constant(HAT_EF_ELE_WATER_1);
+	export_constant(HAT_EF_ELE_EARTH_1);
+	export_constant(HAT_EF_ELE_FIRE_1);
+	export_constant(HAT_EF_ELE_WIND_1);
+	export_constant(HAT_EF_ELE_POISON_1);
+	export_constant(HAT_EF_ELE_HOLY_1);
+	export_constant(HAT_EF_ELE_DARK_1);
+	export_constant(HAT_EF_ELE_GHOST_1);
+	export_constant(HAT_EF_ELE_UNDEAD_1);
+	export_constant(HAT_EF_ELE_NEUTRAL_2);
+	export_constant(HAT_EF_ELE_WATER_2);
+	export_constant(HAT_EF_ELE_EARTH_2);
+	export_constant(HAT_EF_ELE_FIRE_2);
+	export_constant(HAT_EF_ELE_WIND_2);
+	export_constant(HAT_EF_ELE_POISON_2);
+	export_constant(HAT_EF_ELE_HOLY_2);
+	export_constant(HAT_EF_ELE_DARK_2);
+	export_constant(HAT_EF_ELE_GHOST_2);
+	export_constant(HAT_EF_ELE_UNDEAD_2);
+	export_constant(HAT_EF_ELE_NEUTRAL_3);
+	export_constant(HAT_EF_ELE_WATER_3);
+	export_constant(HAT_EF_ELE_EARTH_3);
+	export_constant(HAT_EF_ELE_FIRE_3);
+	export_constant(HAT_EF_ELE_WIND_3);
+	export_constant(HAT_EF_ELE_POISON_3);
+	export_constant(HAT_EF_ELE_HOLY_3);
+	export_constant(HAT_EF_ELE_DARK_3);
+	export_constant(HAT_EF_ELE_GHOST_3);
+	export_constant(HAT_EF_ELE_UNDEAD_3);
+	// [RomuloSM]: Race
+	export_constant(HAT_EF_RC_FORMLESS_1),
+	export_constant(HAT_EF_RC_UNDEAD_1),
+	export_constant(HAT_EF_RC_BRUTE_1),
+	export_constant(HAT_EF_RC_PLANT_1),
+	export_constant(HAT_EF_RC_INSECT_1),
+	export_constant(HAT_EF_RC_FISH_1),
+	export_constant(HAT_EF_RC_DEMON_1),
+	export_constant(HAT_EF_RC_DEMIHUMAN_1),
+	export_constant(HAT_EF_RC_ANGEL_1),
+	export_constant(HAT_EF_RC_DRAGON_1),
+	export_constant(HAT_EF_RC_PLAYER_HUMAN_1),
+	export_constant(HAT_EF_RC_PLAYER_DORAM_1),
+	export_constant(HAT_EF_RC_FORMLESS_2),
+	export_constant(HAT_EF_RC_UNDEAD_2),
+	export_constant(HAT_EF_RC_BRUTE_2),
+	export_constant(HAT_EF_RC_PLANT_2),
+	export_constant(HAT_EF_RC_INSECT_2),
+	export_constant(HAT_EF_RC_FISH_2),
+	export_constant(HAT_EF_RC_DEMON_2),
+	export_constant(HAT_EF_RC_DEMIHUMAN_2),
+	export_constant(HAT_EF_RC_ANGEL_2),
+	export_constant(HAT_EF_RC_DRAGON_2),
+	export_constant(HAT_EF_RC_PLAYER_HUMAN_2),
+	export_constant(HAT_EF_RC_PLAYER_DORAM_2),
+	export_constant(HAT_EF_RC_FORMLESS_3),
+	export_constant(HAT_EF_RC_UNDEAD_3),
+	export_constant(HAT_EF_RC_BRUTE_3),
+	export_constant(HAT_EF_RC_PLANT_3),
+	export_constant(HAT_EF_RC_INSECT_3),
+	export_constant(HAT_EF_RC_FISH_3),
+	export_constant(HAT_EF_RC_DEMON_3),
+	export_constant(HAT_EF_RC_DEMIHUMAN_3),
+	export_constant(HAT_EF_RC_ANGEL_3),
+	export_constant(HAT_EF_RC_DRAGON_3),
+	export_constant(HAT_EF_RC_PLAYER_HUMAN_3),
+	export_constant(HAT_EF_RC_PLAYER_DORAM_3),
+	// [RomuloSM]: Quest
+	export_constant(HAT_EF_QUEST_OBJECTIVE_1);
+	export_constant(HAT_EF_QUEST_OBJECTIVE_2);
+	export_constant(HAT_EF_QUEST_OBJECTIVE_3);
+
 	#undef export_constant
 	#undef export_constant2
 	#undef export_parameter
